### DISCUSSION TITLE
Not able to query single collection on GraphQL with anything other than ID

### DISCUSSION THREAD

**Discussion by creative-andrew** [payloadcms/payload#3178] [Feature Requests & Ideas]
### To Reproduce

It is a common practice to use `slugs` or something similar to identify a collection as an addition to the regular ID.
Right now you can only pass the `id` to GraphQL to get a single collection document.
Hence, you cannot create a `GetPageBySlug` query. 

![image](https://github.com/payloadcms/payload/assets/18492002/398ff95f-d8e7-4e9c-a181-62c2b58613c5)

This makes things really hard in the front-end as I cannot query things by using just their slug and I am forced to pass the full id. 

Payload provides [a way to declare a custom ID](https://payloadcms.com/docs/fields/overview#customizable-id).
However, this won't work in more than one language, as you cannot localize a custom id.

The idea here is to be able to declare a field to be used in GraphQL as a unique identifier other than the ID.
Collections support having unique fields, which would work really nicely with this as they are unique within the collection. However, I am not sure if they are also unique also on each language. 

```ts
export const Pages: CollectionConfig = {
  slug: "pages",
  fields: [
    {
      name: "slug",
      type: "text",
      required: true,
      localized: true,
      unique: true,
      useAsIdentifier: true
    },
};
```
```graphql
{
  Page(id: "mySlug" idType: SLUG)
}
```

`idType` is an enum that by default includes `Id` and other unique identifiers defined in a field. 

Here is an example from WordPress WPGraphQL solving this issue:
https://www.wpgraphql.com/docs/posts-and-pages#single-post-by-slug

### Describe the Bug

Not able to query single collection on GraphQL with anything other than ID

### Payload Version

1.9.0

**Comment by JarrodMFlesch** [comment]
Do you need the slug to be localized? Is that common practice?

**Comment by creative-andrew** [comment]
That is in fact a common practice.
Imagine you have a page with a slug "example-api-development-payload" for EN.
The translation of that page should have its own slug "ejemplo-api-desarrollo-payload" for ES
This is important for SEO.

Here are some webs with that behavior:

https://www.whitehouse.gov/es/invertir/ -> ES
https://www.whitehouse.gov/invest/ -> EN

https://es.wikipedia.org/wiki/Estados_Unidos -> ES
https://en.wikipedia.org/wiki/United_States -> EN

https://www.godaddy.com/domains/domain-broker -> EN
https://www.godaddy.com/es-es/dominios/broker-dominios -> ES

https://www.hauraton.com/en/products/ -> EN
https://www.hauraton.com/fr/produits/ -> FR

Please use the language switcher to see the slug changing.

**Comment by JarrodMFlesch** [comment]
You could use a where query to fetch a single doc though. Yes it will come back in an array of docs, but you will only have 1 document inside. That is normally what we do when we need to fetch documents by slugs, which I know is a common thing.

**Comment by creative-andrew** [comment]
> You could use a where query to fetch a single doc though. Yes it will come back in an array of docs, but you will only have 1 document inside. That is normally what we do when we need to fetch documents by slugs, which I know is a common thing.

Yes, what you suggested is what we ended up doing! However, I think we should leave the issue open until we found a more permanent solution.

**Comment by JarrodMFlesch** [comment]
@creative-andrew, I don't know. I feel like this is a feature request and not an issue? Maybe we should convert it to a discussion. I see the value in it, I just am not sure we will keep it open as an "issue".

**Comment by creative-andrew** [comment]
Sure thing. Let's do that.

**Comment by DanRibbens** [comment]
This is a feature idea that has come up a few times. I propose we add a new property to collections called `useAsSlug`. Which would take a string of a field name (top level only) so that you can say I have this field and I want to use it in my routes instead of the id.

Internally this would be exactly how the current user Preferences endpoints are written and this behavior can be handled internally using the new `useAsSlug` property.

The change will affect the REST routes and GraphQL so that querying can be done by the slug value. It also needs to handle localization so that the query value matches on the locale given to return the correct localized document.

Indexes should be automatically created for `useAsSlug` fields to increase query performance.

The id field will still be present to support relationships, but routing by id will not be possible as the paths will conflict between ids and the slug field.

`Default behavior` /api/posts/:id
`useAsSlug: "title"` /api/posts/my-post

This will affect the admin UI as the PATCH request will need to hit the old value path route when saving document edits.
The DELETE will also need to be changed to hit the path of the `useAsTitle` value with the locale in the request.

**Reply by ssyberg** [reply]
Love this!
